doc/ant-xgboost_user_guide.md

```diff
@@ -0,0 +1,265 @@
+* `n > 0` and `c = 1.0`
+  
+   This reduces to the standard early stopping strategy that is employed by dmlc/xgboost.
+
+* `n = 0` and `c in [0, 1]`
+
+    For example, `n = 0` and `c = 0.8`. This means there should be at least 20% of points after the best metric value. Smaller value in `c` leads to a more conservative convergence test. This rule tests convergence in an adaptive way; for some problem the metric values are noisy and grow slowly, this rule will have better chance to find the optimal model. 
+
+In addition, convergence test understands the optimization direction for all built-in metrics, so there is no need to set `maximize` parameter (defaults to `false`, forgetting to set this parameter often leads to strange behavior when metric value should be maximized).
+
+
+### AutoTrain
+With convergence test, we implement a simple `auto_train` method. There are several components in `auto_train`:
+* Automatic parameter validation, setting and rewriting
+
+    Setting the right parameters in XGBoost is not easy. For example, when working with `binary:logistic`, one should not set `num_class` to 2 (otherwise XGBoost will fail with an exception). 
+    
+    In Ant-XGBoost, we validate parameters to make sure all parameters are consistent with each other, e.g., `num_class = 3` and `objective = binary:logistic` will raise an exception. 
+    
+    In addition, we try our best to understand the input parameters from the user and automatically set or rewrite some parameters in `auto_train` mode. 
+    For example, when the feature dimension is very high, building a single tree will be very inefficient,
+    we will automatically set `colsample_bytree` and make sure at most 2000 features are used to build for each tree. Note that automatic parameter rewritting is only turned on in `auto_train` mode, in standard `train`, we only validate parameters and the behavior is fully controled by the user.
+
+* automatic training
+
+    With convergence test, number of trees in a boosted ensemble becomes a less important parameter; 
+    one can always set a very large number and rely on convergence test to figure out the right number of trees. 
+    The most important parameters to tune in boosted trees now become learning rate and max depth. 
+    In Ant-XGBoost, we employ grid search with early stopping to efficiently search the best model structure; 
+    unpromising learning rate or depth will be skipped entirely.
+    
+    While the current `auto_train` method is a very simple approach, we are working on better strategies to further scale up hyperparameter tuning in XGBoost training.
+
+
+### Helpful Backports of XGBoost master
+We keep watching the progress of XGBoost community.
+Our stable version is based on latest stable version of dmlc/xgboost.
+What's more, we will cherry-pick helpful cutting edge commits, such as blocking issue fix and non-trivial performance improvement.
+
+<br>
+
+## Quick Start
```

Comment:
This document is supposed to be about the design of `antxgboost_codegen.go`. But there is no discussion about this code generator?